### PR TITLE
HTTPC-187: Expression based response validator

### PR DESCRIPTION
**Pull request by eze210**

### CODE REVIEW

File: src/main/java/org/mule/extension/http/api/request/validator/ExpressionResponseValidator.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+import org.mule.runtime.http.api.domain.message.request.HttpRequest;
+
+import javax.inject.Inject;
+import java.io.InputStream;
+import java.util.Optional;
+
+/**
+ * Response validator that uses an expression.
+ *
+ * @since 1.0
```
Comment (mbuchwald) [inline src/main/java/org/mule/extension/http/api/request/validator/ExpressionResponseValidator.java:33]:
```suggestion
 * @since 1.7
```
?

**Review by mbuchwald — CHANGES_REQUESTED** [review]

**Review by mbuchwald — APPROVED** [review]

File: src/main/java/org/mule/extension/http/api/request/validator/ExpressionResponseValidator.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+  @Inject
+  private ExpressionManager expressionManager;
+
+  @Override
+  public void validate(Result<InputStream, HttpResponseAttributes> result, HttpRequest request) {
+    Optional<String> optExpression = expression.getLiteralValue();
+    if (!optExpression.isPresent()) {
+      throw new ResponseValidatorTypedException("An expression is required", BAD_REQUEST);
+    }
```
Comment (elrodro83) [inline src/main/java/org/mule/extension/http/api/request/validator/ExpressionResponseValidator.java:53]: isn't this already covered with `expression` not being `@Optional`?
Also, this kind of validations should be done just once, and not for revery request.

File: src/main/java/org/mule/extension/http/internal/request/HttpRequester.java [hunk 3]
```diff
@@ -221,6 +223,12 @@ private void doRequestWithRetry(HttpExtensionClient client, HttpRequesterConfig
     return () -> (InputStream) resolved;
   }
 
+  private static boolean bodyCanBeConsumed(HttpRequestAuthentication authentication, ResponseValidator responseValidator) {
+    boolean authCanConsumeBody = authentication != null && authentication.readsAuthenticatedResponseBody();
+    boolean validatorCanConsumeBody = responseValidator != null && responseValidator.mayConsumeBody();
+    return authCanConsumeBody || validatorCanConsumeBody;
+  }
```
Comment (elrodro83) [inline src/main/java/org/mule/extension/http/internal/request/HttpRequester.java:230]: can -> may

File: src/test/java/org/mule/extension/http/api/request/validator/ExpressionResponseValidatorTestCase.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+import org.mule.runtime.extension.api.runtime.parameter.Literal;
+import org.junit.Test;
+import org.mule.functional.api.exception.ExpectedError;
+import org.mule.runtime.http.api.domain.message.request.HttpRequest;
+import org.mule.test.http.functional.requester.AbstractHttpRequestTestCase;
+
+import java.io.InputStream;
+import java.util.Optional;
+
+public class ExpressionResponseValidatorTestCase extends AbstractHttpRequestTestCase {
```
Comment (elrodro83) [inline src/test/java/org/mule/extension/http/api/request/validator/ExpressionResponseValidatorTestCase.java:33]: allure

**Review by elrodro83 — CHANGES_REQUESTED** [review]

File: src/main/java/org/mule/extension/http/api/request/validator/ExpressionResponseValidator.java (hunk 2 above)
Comment (eze210) [reply src/main/java/org/mule/extension/http/api/request/validator/ExpressionResponseValidator.java]: Yes, but the Literal interface returns an Optional<>, so the most correct option is to check if the value isPresent.
Adding a comment and throwing an IllegalStateException instead.

File: src/main/java/org/mule/extension/http/internal/request/HttpRequester.java (hunk 3 above)
Comment (eze210) [reply src/main/java/org/mule/extension/http/internal/request/HttpRequester.java]: Done

File: src/test/java/org/mule/extension/http/api/request/validator/ExpressionResponseValidatorTestCase.java (hunk 4 above)
Comment (eze210) [reply src/test/java/org/mule/extension/http/api/request/validator/ExpressionResponseValidatorTestCase.java:39]: Done

**Review by elrodro83 — APPROVED** [review]

**Review by marianogonzalez — CHANGES_REQUESTED** [review]

File: src/test/java/org/mule/extension/http/api/request/validator/ExpressionResponseValidatorTestCase.java [hunk 5]
```diff
@@ -75,14 +76,21 @@ public void expressionEvaluatingToFalse() throws Exception {
   }
 
   @Test(expected = IllegalStateException.class)
-  public void whenLiteralExpressionIsNotPresentTheValidatorThrowsException() {
+  public void whenLiteralExpressionIsNotPresentTheValidatorThrowsException() throws NoSuchFieldException, IllegalAccessException {
```
Comment (marianogonzalez) [inline src/test/java/org/mule/extension/http/api/request/validator/ExpressionResponseValidatorTestCase.java:79]: why isn't this an MTF test?

File: src/test/java/org/mule/extension/http/api/request/validator/ExpressionResponseValidatorTestCase.java [hunk 6]
```diff
@@ -75,14 +76,21 @@ public void expressionEvaluatingToFalse() throws Exception {
...
     ExpressionResponseValidator validator = new ExpressionResponseValidator();
     Literal<String> mockLiteral = mock(Literal.class);
     when(mockLiteral.getLiteralValue()).thenReturn(empty());
-    validator.setExpression(mockLiteral);
+    setExpression(validator, mockLiteral);
```
Comment (marianogonzalez) [inline src/test/java/org/mule/extension/http/api/request/validator/ExpressionResponseValidatorTestCase.java:83]: ideally you should do this with MTF and not require setter at all. But it you have no choice than to do this, then leave the setter but with package scope

**Review by marianogonzalez — CHANGES_REQUESTED** [review]

**Pull request merged**